src/cryptonote_protocol/cryptonote_protocol_handler.inl

```diff
@@ -2532,14 +2532,21 @@ skip:
 
     context.m_needed_objects.clear();
     uint64_t added = 0;
+    std::unordered_set<crypto::hash> blocks_found;
     for (size_t i = 0; i < arg.m_block_ids.size(); ++i)
     {
+      blocks_found.insert(arg.m_block_ids[i]);
       const uint64_t block_weight = arg.m_block_weights.empty() ? 0 : arg.m_block_weights[i];
       context.m_needed_objects.push_back(std::make_pair(arg.m_block_ids[i], block_weight));
       if (++added == n_use_blocks)
         break;
     }
-    context.m_last_response_height -= arg.m_block_ids.size() - n_use_blocks;
+    if (blocks_found.size() != arg.m_block_ids.size())
```

Comment:
This assumes that the previous loop always iterates without early exit. Are you sure `n_use_blocks` is always the same as `arg.m_block_ids.size()`?

Reply:
I think it's safer to check the return value of `blocks_found.insert(arg.m_block_ids[i])` to see if it actually inserted new element or not. From C++ docs:
>  the function returns a pair object whose first element is an iterator pointing either to the newly inserted element in the container or to the element whose key is equivalent, and a bool value indicating whether the element was successfully inserted or not.